extensions/native/recursion/src/stark/mod.rs

```diff
@@ -537,27 +557,38 @@ where
             .for_each(|phase_idx_vec, builder| {
                 let phase_idx = phase_idx_vec[0];
                 let values_per_mat = builder.get(&opening.values.after_challenge, phase_idx);
+                builder.assert_usize_eq(values_per_mat.len(), num_after_challenge_traces);
                 let batch_commit = builder.get(after_challenge_commits, phase_idx);
 
                 let mat_idx: Usize<_> = builder.eval(RVar::zero());
-                let mats: Array<_, TwoAdicPcsMatsVariable<_>> = builder.array(values_per_mat.len());
+                let mats: Array<_, TwoAdicPcsMatsVariable<_>> =
+                    builder.array(num_after_challenge_traces);
                 builder.range(0, num_airs).for_each(|i_vec, builder| {
                     let i = i_vec[0];
                     let advice = builder.get(&m_advice_var.per_air, i);
                     builder
-                        .if_ne(advice.num_challenges_to_sample.len(), RVar::zero())
+                        .if_ne(advice.width.after_challenge.len(), RVar::zero())
                         .then(|builder| {
                             // Only 1 phase is supported.
-                            builder.assert_usize_eq(
-                                advice.num_challenges_to_sample.len(),
-                                RVar::one(),
+                            builder
+                                .assert_usize_eq(advice.width.after_challenge.len(), RVar::one());
+                            let after_challenge_width = RVar::from(
+                                builder.get(&advice.width.after_challenge, RVar::zero()),
                             );
                             let domain = builder.get(&domains, i);
                             let trace_points = builder.get(&trace_points_per_domain, i);
 
                             let after_challenge = builder.get(&values_per_mat, mat_idx.clone());
 
                             let values = builder.array::<Array<C, _>>(2);
+                            builder.assert_usize_eq(
+                                after_challenge.local.len(),
+                                after_challenge_width * RVar::from(C::EF::D),
+                            );
+                            builder.assert_usize_eq(
+                                after_challenge.next.len(),
+                                after_challenge_width * RVar::from(C::EF::D),
```

Comment:
nit: It should be more efficient to compare to `after_challenge.local.len()` here.